src/Io/Transaction.php

```diff
@@ -255,25 +258,37 @@ private function onResponseRedirect(ResponseInterface $response, RequestInterfac
     /**
      * @param RequestInterface $request
      * @param UriInterface $location
+     * @param int $statusCode
      * @return RequestInterface
+     * @throws \RuntimeException
      */
-    private function makeRedirectRequest(RequestInterface $request, UriInterface $location)
+    private function makeRedirectRequest(RequestInterface $request, UriInterface $location, $statusCode)
     {
         $originalHost = $request->getUri()->getHost();
-        $request = $request
-            ->withoutHeader('Host')
-            ->withoutHeader('Content-Type')
-            ->withoutHeader('Content-Length');
+        $request = $request->withoutHeader('Host')->withUri($location);
+
+        if (!$this->isTemporaryOrPermanentRedirect($statusCode)) {
+            $request = $request
+                ->withoutHeader('Content-Type')
+                ->withoutHeader('Content-Length');
+        }
 
         // Remove authorization if changing hostnames (but not if just changing ports or protocols).
         if ($location->getHost() !== $originalHost) {
             $request = $request->withoutHeader('Authorization');
         }
 
-        // naïve approach..
-        $method = ($request->getMethod() === 'HEAD') ? 'HEAD' : 'GET';
+        if ($this->isTemporaryOrPermanentRedirect($statusCode)) {
+            if ($request->getBody() instanceof ReadableStreamInterface) {
+                throw new \RuntimeException('Unable to redirect request with streaming body');
+            }
+        } else {
+            // naïve approach..
+            $method = ($request->getMethod() === 'HEAD') ? 'HEAD' : 'GET';
+            $request = $request->withMethod($method)->withBody(new EmptyBodyStream());
+        }
```

Comment:
```suggestion
        if ($statusCode === Response::STATUS_TEMPORARY_REDIRECT || $statusCode === Response::STATUS_PERMANENT_REDIRECT) {
            if ($request->getBody() instanceof ReadableStreamInterface) {
                throw new \RuntimeException('Unable to redirect request with streaming body');
            }
        } else {
            $request = $request
                ->withMethod($request->getMethod() === 'HEAD' ? 'HEAD' : 'GET')
                ->withoutHeader('Content-Type')
                ->withoutHeader('Content-Length')
                ->withBody(new EmptyBodyStream());
        }
```
Part B2: This is the result from B1, I also optimized the code a bit. This way we don't need the extra `isTemporaryOrPermanentRedirect($statusCode)` function. (**See part B3**)